### DISCUSSION TITLE
Partitioned Workspace vs Poled Workspace?

### DISCUSSION THREAD

**Discussion by maxim-ge** [voedger/voedger#102] [Polls]
For Edge Devices Registry we need to keep Edge Devices states (~ 10.000 devices).

1. `One workspace for all devices` is a non-scalable bottleneck
1. `One workspace per device` is ok for keeping status, but not ok for quering status, sending 10.000 queries is not ok.

Possible solution would be to have 10-100 EdgeDevicesStatus workspaces inside EdgeDevicesRegistry workspace, it is ok to send 10-100 queries and merge results.

Below are syntax options to define EdgeDevicesRegistry - EdgeDevicesStates relation.

### POOL OF WORKSPACE

```sql

WORKSPACE EdgeDevicesRegistry (
  POOL OF WORKSPACE EdgeDevicesStates (
     Def1
     Def2
     Def3
  )

)
```

### PARTITIONED WORKSPACE

```sql

WORKSPACE EdgeDevicesRegistry (
  PARTITIONED WORKSPACE EdgeDevicesStates (
     Def1
     Def2
     Def3
  )

)
```

### Attribute PARTITIONED 

```sql

WORKSPACE EdgeDevicesRegistry (
  WORKSPACE EdgeDevicesStates (
    PARTITIONED
    Def1
    Def2
    Def3
  )

)
```

**Comment by maxim-ge** [comment]
Note that we have a very similar situation with the Cloud Bill product - we need to keep bills from thousands of restaurants, calculate and query aggregates.

**Comment by host6** [comment]
have 100.000 devices -> 100.000 queries for statuses in a row -> storage is overloaded by reading. Correct?

**Reply by maxim-ge** [reply]
> have 100.000 devices -> 100.000 queries for statuses in a row -> storage is overloaded by reading. Correct?

Data will be taken from the cache, so storage won't be overloaded. I think it will take at least ~ 100 seconds to send 100.000 queries, that's the main problem.

**Comment by AlekseiPonomarev** [comment]
May be:
```
WORKSPACE EdgeDevicesRegistry (
  ...
  WORKSPACE EdgeDevicesStates PARTITION BY (
    ...
  )

)
```

**Reply by maxim-ge** [reply]
>   WORKSPACE EdgeDevicesStates PARTITION BY [have a feeling that something is missed here] (

**Comment by maxim-ge** [comment]
Another situation -  100.000 sensors, each sensor generates one event per second. Each workspace can handle ~ 10000 per second.

Therefore we need a pool of at least 10 workspaces (= 10 partitions).

**Reply by host6** [reply]
do not understand what is "pool" here. Each of 100.000 sensors is a separate workspace already, all 100.000 workspaces are spread over 10 partitions -> the entire cluster has a constant high load (no overload)

**Reply by maxim-ge** [reply]
> Each of 100.000 sensors is a separate workspace already

No, it is suggested to use a pool of the 10-100 workspaces to keep data from sensors, so 1.000-10.000 sensors per workspace.

You can also think of it as a single PARTITIONED workspace.

**Comment by michael-say** [comment]
By the way, why do not specify pool size directly in statement?
```sql
POOL OF 100 WORKSPACE EdgeDevicesStates (
...
)
```

**Reply by maxim-ge** [reply]
> POOL OF 100 WORKSPACE EdgeDevicesStates (

Good question, I had this in mind but it is hard to initialize the Pool - workspaces have to be created in background during so called `Recovery` procedure which can take a while for development. To deliver solution in a short term we will add workspaces to the Pool one-by-one using API.
